crates/hyperdrive-math/src/lp.rs

```diff
@@ -600,6 +621,567 @@ impl State {
+        &self,
+        active_lp_total_supply: FixedPoint, // just the total number of lp shares
+        starting_present_value: FixedPoint,
+        lp_total_supply: FixedPoint, // total lp shares and withdrawal shares - w.s. ready to withraw
+        present_value: FixedPoint,
+    ) -> Result<bool> {
+        if active_lp_total_supply == fixed!(0) {
+            return Err(eyre!("active_lp_total_supply is zero"));
+        }
+        if lp_total_supply == fixed!(0) {
+            return Err(eyre!("lp_total_supply is zero"));
+        }
+        let new_lp_share_price_down = present_value.div_down(active_lp_total_supply);
+        let new_lp_share_price_up = present_value.div_up(active_lp_total_supply);
+        let old_lp_share_price = starting_present_value.div_up(lp_total_supply);
+        let old_lp_share_price_with_tolerance = (fixed!(1e18)
+            + FixedPoint::from(SHARE_PROCEEDS_SHORT_CIRCUIT_TOLERANCE))
+        .mul_div_down(starting_present_value, lp_total_supply);
+
+        Ok(
+            // Ensure that new LP share price is greater than or equal to the
+            // previous LP share price:
+            //
+            // PV_1 / l_1 >= PV_0 / l_0
+            //
+            // NOTE: Round the LHS down to make the check stricter.
+            new_lp_share_price_down >= old_lp_share_price
+            // Ensure that new LP share price is less than or equal to the
+            // previous LP share price plus the minimum tolerance:
+            //
+            // PV_1 / l_1 <= (PV_0 / l_0) * (1 + tolerance)
+            //
+            // NOTE: Round the LHS up to make the check stricter.
+            && new_lp_share_price_up <= old_lp_share_price_with_tolerance,
+        )
+    }
+
+    /// Calculates the amount of withdrawal shares that can be redeemed and
+    /// the share proceeds the withdrawal pool should receive given the
+    /// pool's current idle liquidity. We use the following algorithm to
+    /// ensure that the withdrawal pool receives the correct amount of
+    /// shares to (1) preserve the LP share price and (2) pay out as much
+    /// of the idle liquidity as possible to the withdrawal pool:
+    ///
+    /// 1. If `y_s * t_s <= y_l * t_l` or
+    ///    `y_max_out(I) >= y_s * t_s - y_l * t_l`, set `dz_max = I` and
+    ///    proceed to step (3). Otherwise, proceed to step (2).
+    /// 2. Solve `y_max_out(dz_max) = y_s * t_s - y_l * t_l` for `dz_max`
+    ///    using Newton's method.
+    /// 3. Set `dw = (1 - PV(dz_max) / PV(0)) * l`. If `dw <= w`, then
+    ///    proceed to step (5). Otherwise, set `dw = w` and continue to
+    ///    step (4).
+    /// 4. Solve `PV(0) / l = PV(dz) / (l - dw)` for `dz` using Newton's
+    ///    method if `y_l * t_l != y_s * t_s` or directly otherwise.
+    /// 5. Return `dw` and `dz`.
+    ///
+    ///  Returns (withdrawal_shares_redeemed, share_proceeds)
+    pub fn calculate_distribute_excess_idle(
```

Comment:
This is meta-comment that applies to a bunch of lines. I'm happy to clarify which ones I'm talking about if you aren't sure about some cases. 

In Solidity, we use the pattern:

```
(uint256 X, bool success) = calculateBlahBlahBlahSafe(someArgs);
if (!success) {
    return (0, false);
}
```

in a bunch of cases. This is roughly equivalent to the `?` operator in Rust:

```
let X = calculate_blah_blah_blah_safe(some_args)?;
```

Where this breaks down is in `calculate_distribute_excess_idle`. In Solidity, `calculate_distribute_excess_idle` is NOT a safe function, so it doesn't return a failure flag. The implication of this is that `calculate_distribute_excess_idle` doesn't just unwrap all of the underlying results (this is an analogy, but stick with me for a sec) using `?`. Instead, it returns `0`  when a failure flag is returned. Here's an example of that: https://github.com/delvtech/hyperdrive/blob/a98362d1ead7c068957d1bdcfed9c6d7b0c7e4e8/contracts/src/libraries/LPMath.sol#L587.

With this in mind, there are a bunch of errors that this implementation will treat as failures that Solidity would handle as "no idle can be distributed."

Reply:
Ok I went through the PR and it much more closely aligns with the solidity now!